Sources/SwiftCompilerPlugin/CompilerPlugin.swift

```diff
@@ -281,3 +303,24 @@ private func describe(errno: CInt) -> String {
   if let cStr = strerror(errno) { return String(cString: cStr) }
   return String(describing: errno)
 }
+
+#if os(WASI)
+
+/// A callback invoked by the Wasm Host when new data is available on `stdin`.
+private var readabilityHandler: () -> Void = {
+  internalError("""
+  CompilerPlugin.main wasn't called. Did you annotate your plugin with '@main'?
+  """)
+}
+
+#if compiler(>=6.0)
+
+@_expose(wasm, "swift_wasm_macro_pump")
+@_cdecl("swift_wasm_macro_pump")
+func wasmPump() {
+  readabilityHandler()
+}
+
+#endif
```

Comment:
Does this code require Swift 6 to build for wasm? What happens if you build using a Swift 5 compiler? If it doesn’t build with an older compiler, I think we should put a `#error` in the `#else` branch telling you that you need a Swift 6 compiler to build the plugin server for WASM.

Reply:
Turns out this was causing pre-6.0 builds (even those not targeting wasm) to fail — somehow Swift ignores the `#if compiler` check and complains that it doesn't recognize `@_expose(wasm)`. I moved the export over to C, so this should now work fine with older compilers. That said I haven't tested a pre-6.0 Wasm-compatible compiler yet, since afaik Swift didn't officially support Wasm until 6.0 — but I can give it a spin with a SwiftWasm toolchain.

Reply:
FWIW, if you're having trouble with an `#if compiler` check, use something like `#if compiler(>=5.3) && hasAttribute(_expose)`, and older compilers won't look into the `#if` at all.

Reply:
So this didn't quite work but it got me poking around the `Parser::parseIfConfig` code and I discovered a bug:

The following fails to compile
```swift
#if os(WASI)
#if compiler(>=100)
foo
#endif
#endif
```

whereas it compiles fine if we swap the conditions
```swift
#if compiler(>=100)
#if os(WASI)
foo
#endif
#endif
```

Seems like if we're already in an inactive `#if` clause, the parser doesn't respect the "disable parsing entirely" functionality of nested `#if compiler` decls. I'll see if I can make a followup in apple/swift to fix this but in the meantime the fix is thankfully as easy as swapping the statements around.

Reply:
https://github.com/apple/swift/pull/74415